src/librustc_resolve/lib.rs

```diff
@@ -3457,6 +3502,86 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
+        let mut lookup_modules = Vec::new();
+        lookup_modules.push((self.graph_root, Vec::new()));
+
+        while let Some((in_module, path_segments)) = lookup_modules.pop() {
+            build_reduced_graph::populate_module_if_necessary(self, &in_module);
+
+            in_module.for_each_child(|name, ns, name_binding| {
+
+                // avoid imports entirely
+                if !name_binding.is_import() {
+
+                    // collect results based on the filter function
+                    if let Some(def) = name_binding.def() {
+                        if ns == namespace {
+                            if filter_fn(def) && name == lookup_name {
+                                // create the path
+                                let ident = hir::Ident::from_name(name);
+                                let params = PathParameters::none();
+                                let segment = PathSegment {
+                                    identifier: ident,
+                                    parameters: params,
+                                };
+                                let span = name_binding.span.unwrap_or(
+                                    syntax::codemap::DUMMY_SP);
+                                let mut segms = path_segments.clone();
+                                segms.push(segment);
+                                let segms = HirVec::from_vec(segms);
+                                let path = Path {
+                                    span: span,
+                                    global: true,
+                                    segments: segms,
+                                };
+                                // push result
+                                lookup_results.push(path);
+                            }
+                        }
+                    }
+
+                    // collect submodules to explore
+                    if let Some(module) = name_binding.module() {
+                        // form the path
+                        let path_segments =
+                            match module.parent_link {
+                                NoParentLink =>
+                                    path_segments.clone(),
+                                ModuleParentLink(_, name) => {
+                                    let mut paths = path_segments.clone();
+                                    let ident = hir::Ident::from_name(name);
+                                    let params = PathParameters::none();
+                                    let segm = PathSegment{
+                                        identifier: ident,
+                                        parameters: params,
+                                    };
+                                    paths.push(segm);
+                                    paths
+                                }
+                                _ =>
+                                    unreachable!()
```

Comment:
Since `unreachable!()` isn't in a block, it's most idiomatic to have it on the same line with a trailing comma:
`_ => unreachable!(),`